### ISSUE TITLE
Throttle option added to the watch

### ISSUE THREAD

**Issue by YeungKC** [simolus3/drift#1510]
Sometimes we don't always need to watch the local database in real time, for some less important data, we can use the throttle to control it completely.

I think we should have a default setting for this, or export a class like StreamQueryStore.

**Comment by YeungKC** [comment]
About this change，the cpu usage become very hight on our project. should be related to stream.

**Comment by simolus3** [comment]
I didn't release the stream optimizations yet, are you using a version from git? I tried to reproduce it with the following program and it didn't consume lots of CPU

```dart
Future<void> main() async {
  final db = Database(NativeDatabase.memory());
  var events = 0;

  final sub = db
      .select(db.categories)
      .watch()
      .throttled(const Duration(milliseconds: 50))
      .listen((_) => events++);

  for (var i = 0; i < 1000; i++) {
    db.markTablesUpdated({db.categories});
    await Future.delayed(const Duration(milliseconds: 10));
  }

  print('Emitted $events times.');
  await sub.cancel();
}
```

If you run something similar with your database class, what does it print for you? The number shouldn't be much larger than 200.

**Comment by YeungKC** [comment]
https://github.com/simolus3/moor/commit/9b273449c308ebb305e3f7cf1d4fd15e2cf8efbb#diff-543f7907aa0e4fb0a36868ab01f19c926887423012e064810a56914ae3373c01R221-R227

I believe it is this code that causes the constant calls to fetchAndEmitData and then pause and then resume, resulting in an infinite loop.

**Comment by simolus3** [comment]
Thanks for looking into it! I don't really see how `fetchAndEmitData` would be called a lot (it definitely shouldn't be called more often just because there are more listeners), and from a quick look the code looks alright (we only call `fetchAndEmitData` if we have an unpaused listener, aren't fetching already and don't have a valid snapshot in the cache).

If you have some simple code that reproduces a high CPU load by interacting with query streams, could you share that with me? Or can you describe what you're doing that causes this (e.g. frequent pause/resumes, cancelling the stream a lot, ...)? If I can reproduce this I'll take a look at fixing this.

**Comment by YeungKC** [comment]
I believe the problem is clear, but you need a demo, so here it is.


https://user-images.githubusercontent.com/11473691/142875834-5632aa60-582b-4ccd-81d1-02720f6a8511.mp4

main.dart:
```dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

import 'database.dart';

Future<void> main() async {
  final myDatabase = MyDatabase();
  runApp(Provider.value(
    value: myDatabase,
    child: const MyApp(),
  ));
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

var updateCount = 0;

class MyHomePage extends StatelessWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) => Scaffold(
        body: StreamBuilder(
          stream: context.read<MyDatabase>().allTodoEntries,
          builder: (BuildContext context, AsyncSnapshot<List<Todo>> snapshot) {
            updateCount += 1;
            return Column(
              children: [
                Text('updateCount: $updateCount'),
                Text(snapshot.data?.toString() ?? 'loading...'),
              ],
            );
          },
        ),
      );
}
```

database.dart:
```dart

import 'dart:async';
import 'dart:io';

import 'package:drift/drift.dart';
import 'package:drift/native.dart';
import 'package:path/path.dart';
import 'package:path_provider/path_provider.dart';

part 'database.g.dart';

class Todos extends Table {
  IntColumn get id => integer().autoIncrement()();

  TextColumn get title => text().withLength(min: 6, max: 32)();

  TextColumn get content => text().named('body')();

  IntColumn get category => integer().nullable()();
}

@DataClassName("Category")
class Categories extends Table {
  IntColumn get id => integer().autoIncrement()();

  TextColumn get description => text()();
}

LazyDatabase _openConnection() {
  return LazyDatabase(() async {
    final dbFolder = await getApplicationDocumentsDirectory();
    final file = File(join(dbFolder.path, 'db.sqlite'));
    return NativeDatabase(file);
  });
}

@DriftDatabase(tables: [Todos, Categories])
class MyDatabase extends _$MyDatabase {
  MyDatabase() : super(_openConnection());

  @override
  int get schemaVersion => 1;

  late final allTodoEntries =
      select(todos).watch().throttled(const Duration(milliseconds: 30));
}

extension ThrottleWithPause<T> on Stream<T> {
  Stream<T> throttled(Duration duration) {
    return Stream.multi(
      (listener) {
        late final StreamSubscription<T> subscription;

        void onEvent() {
          subscription.pause(Future.delayed(duration));
        }

        subscription = listen(
          (event) {
            onEvent();
            listener.addSync(event);
          },
          onError: (Object e, StackTrace s) {
            onEvent();
            listener.addErrorSync(e, s);
          },
          onDone: () {
            subscription.cancel();
            listener.closeSync();
          },
        );

        listener
          ..onPause = subscription.pause
          ..onResume = subscription.resume
          ..onCancel = subscription.cancel;
      },
      isBroadcast: isBroadcast,
    );
  }
}

```

**Comment by davidmartos96** [comment]
@YeungKC You seem to be recreating the stream in every build. Futures and Streams should be handled in stateful widgets by creating them once in the initState

**Comment by YeungKC** [comment]
@davidmartos96 



https://user-images.githubusercontent.com/11473691/142877663-45dbccb8-9086-4ed8-8298-e32a8c778549.mp4



```dart

var updateCount = 0;
var streamBuilderUpdateCount = 0;

class MyHomePage extends StatelessWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) => Scaffold(
        body: Builder(builder: (context) {
          updateCount += 1;

          return Column(
            children: [
              Text('updateCount: $updateCount'),
              StreamBuilder(
                stream: context.read<MyDatabase>().allTodoEntries,
                builder:
                    (BuildContext context, AsyncSnapshot<List<Todo>> snapshot) {
                  streamBuilderUpdateCount += 1;
                  return Column(
                    children: [
                      Text(
                          'streamBuilderUpdateCount: $streamBuilderUpdateCount'),
                      Text(snapshot.data?.toString() ?? 'loading...'),
                    ],
                  );
                },
              ),
            ],
          );
        }),
      );
}


```

**Comment by simolus3** [comment]
Actually even creating the stream in every `build` method shouldn't have much of a performance impact with drift. The `StreamBuilder` will create a new subscription in that case, but that's not a big deal since drift is smart enough to not re-run the query. So the only overhead is that your `StreamBuilder` will update one additional time.

I ran your example and enabled `logStatements: true` for the `NativeDatabase`. Looking at the logs, you can see that drift does _not_ actually re-run the query. It does however, emit the latest cached snapshot to a resuming listener even if it didn't change. I didn't realize this initially because I only looked at how often the query was sent, but we shouldn't emit stale data either of course. 
So thanks for the helpful repro! I have fixed this in 871f1e4198da994060b02a9eaef9a809d5750918.
